src/Microsoft.AspNetCore.SignalR.Core/SerializedHubMessage.cs

```diff
@@ -35,21 +36,55 @@ public SerializedHubMessage(HubMessage message)
 
         public ReadOnlyMemory<byte> GetSerializedMessage(IHubProtocol protocol)
         {
+            // Double-check locking!
             if (!TryGetCached(protocol.Name, out var serialized))
```

Comment:
This will check the list which isn't thread safe. Should only check:

```
             if (string.Equals(_cachedItem1.ProtocolName, protocolName, StringComparison.Ordinal))
             {
                 result = _cachedItem1.Serialized;
                 return true;
             }
 
             if (string.Equals(_cachedItem2.ProtocolName, protocolName, StringComparison.Ordinal))
             {
                 result = _cachedItem2.Serialized;
                 return true;
             }
```

Reply:
Yep, good catch

Reply:
Shoot, I think this could shear... The struct is two fields so it's not guaranteed to set atomically.

Reply:
To clarify (I was headed out the door when I found this): When setting this, since the `_cachedItem` fields are multi-field structs, a thread could set the Protocol Name but **not** yet have set the Serialized message when the reader comes along to check for a cached copy. This would result in returning an empty `ReadOnlyMemory` as the serialized version for a message (which would be A Bad Thing).

Reply:
gotcha

Reply:
We can probably keep the optimization if we use an atomic counter to track the number of items. I'll look at that.

Reply:
Yeah, we just need an atomic way to know that a field has been fully assigned. An `Interlocked.Increment`ed counter should achieve that